docs/sdk-configuration.md

```diff
@@ -0,0 +1,676 @@
+
+MySQL is the only known corner case that requires the global SDK instance. If such a corner case
+didn't exist, we may not even support it in the first place.
+
+See special note about Java Agent below though.
+
+## Telemetry within SDK components
+
+SDK components, such as exporters or remote samplers, may want to emit telemetry for their own
+processing. However, the SDK components must be initialized before the SDK can be fully built. We do
+not support partially built SDK because one cannot reason about the behavior of it. Similarly we
+do not support using the global instance of the SDK before it has been built. Therefore, SDK
+components that require `OpenTelemetry` must accept it lazily. This is a restriction, but given
+such components are rarely developed by application developers, and generally developed by either
+framework authors or OpenTelemetry maintainers, this restriction is deemed reasonable.
+
+If this mechanism was built into the SDK, it may look like
+
+```java
+interface OpenTelemetryComponent {
+  default void setOpenTelemetry(OpenTelemetry openTelemetry) {}
+}
+interface SpanExporter extends OpenTelemetryComponent {
+}
+public class BatchExporter implements SpanExporter {
+  
+  private volatile Tracer tracer;
+  
+  @Override
+  public void setOpenTelemetry(OpenTelemetry openTelemetry) {
+        tracer = openTelemetry.getTracerProvider().get("spanexporter");
+  }
+  
+  @Override
+  public void export() {
+        Tracer tracer = this.tracer;
+        if (tracer != null) {
+          tracer.spanBuilder("export").startSpan();
+        }
+  }
+}
+public class OpenTelemetrySdkBuilder {
+    public OpenTelemetrySdkBuilder addSpanExporter(SpanExporter exporter) {
+        tracerProvider.addSpanExporter(exporter);
+        components.add(exporter);
+    }
+    
+    public OpenTelemetrySdkBuilder setSampler(Sampler sampler) {
+        tracerProvider.setSampler(sampler);
+        components.add(sampler);
+    }
+    
+    public OpenTelemetry build() {
+        OpenTelemetrySdk sdk = new OpenTelemetrySdk(tracerProvider.build(), meterProvider.build());
+        for (OpenTelemetryComponent component : components) {
+          component.setOpenTelemetry(sdk);
+        }
+    }
+}
+```
+
+A framework author will have an even easier time since most dependency injection frameworks
+natively support lazy injection.
+
+```java
+@Component
+public class MonitoringModule {
+  
+    @Bean
+    @ForSpanExporter
+    public Tracer tracer(TracerProvider tracerProvider) {
+    return tracerProvider.get("spanexporter");
+    }
+}
+
+@Component
+public class MyExporter implements SpanExporter {
+
+    private Lazy<Tracer> tracer;
+    
+    @Inject
+    public MyExporter(@ForSpanExporter Lazy<Tracer> tracer) {
+    this.tracer = tracer;
+    }
+    
+    @Override
+    public void export() {
+    tracer.get().spanBuilder("export").startSpan();
+    }
+}
+```
+
+## Immutability of OpenTelemetry configuration
```

Comment:
Please let me know what's confusing since I want to clarify it :)

Reply:
E.g. I don't understand the connection between `addSpanProcessor` method (on which class, btw?) and telemetry within telemetry (btw, is this the same as what you called "telemetry within SDK components" above?).

"the SDK default implementation is static" - do you mean "immutable"? And in general, sorry Anuraag, this section reads much sloppier than the preceding document :)

Also, I understand that Bogdan suggests having immutable SDK. As this does make total case on the first glance, may be this section better address those cases, when we potentially need mutability. E.g. "why obvious choice of immutable SDK is not that simple?"

Reply:
Thanks will try to rewrite it - one thing that's always tricky for me is talking about immutability. The SDK isn't immutable in the strictest sense since, e.g., `BatchSpanProcessor` has a mutable queue of spans. I guess the word to use is shallow-immutable which is what I meant by configuration being static. Let me see what I can do.